Duotone: Output duotone presets per block

scruffian opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
This is another angle on https://github.com/WordPress/gutenberg/pull/48291.

If we remove the code that outputs the presets, how can we still use presets? This PR suggests saving the presets to an array and then outputting them in the footer. We do this for both block supports, and filters that are defined in the theme.json.

## Why?
Outputting unused presets creates extra output that we don't need.

## How?
1. Create a new array called $duotone_presets.
2. When we render a block we check if the block has a duotone filter applied - if it does we save it to the new array.
3. We also check if the block has any filter settings in the theme.json - if it does then we also save this setting to the new array.
4. We then run an action on wp_footer which outputs all the used preset duotones.

## Testing Instructions
1. Switch to a theme that has a duotone setting in its theme.json. (eg. Skatepark)
2. Check that the images all have a duotone applied in the front end.
3. Add an image to a post and select a preset duotone filter that is different to the main one for the theme.
4. Check that this image also has a preset duotone filter applied in the frontend.
5. Check that only the used SVGs are output in the frontend

## Notes
Still to do:
- [ ] Find a way to cache the theme json settings so we dont recalculate on every render
- [ ] Combine the two wp_footer actions into one but saving custom duotones in the WP_Duotone class.

scruffian committed:
Duotone: Output presets in block supports

scruffian committed:
remove filters

jeryj commented on lib/block-supports/duotone.php:
```diff
@@ -560,3 +572,44 @@ static function () use ( $filter_svg, $selector ) {
+			 */
+			printf(
+				'<script>( function() { var el = document.querySelector( %s ); var display = el.style.display; el.style.display = "none"; el.offsetHeight; el.style.display = display; } )();</script>',
+				wp_json_encode( $selector )
```
`$selector` isn't defined in this function. Not sure what it's needed for, but it appears to already run through [the footer action above](https://github.com/WordPress/gutenberg/pull/48374/files#diff-fd626ca219f3d805667d480eabb2d890a5a8d2e015b37f03cf8980edf1ae3d58L522-L537) even if it's a global style, so maybe this Safari stuff isn't needed here?

jeryj commented:
I've tested this on TT3 with the Sherbert theme on a single post template with these images:
- Featured image with duotone from global styles
- gutenberg default duotone
- custom selected duotone
- Theme default duotone
- Plain image
- Unset duotone image

It's working in all these cases and only the duotones used on that post are being output in the `<svg>` footer definitions.

jeryj committed:
Check if duotone is unset before adding to duotone presets global

'unset' is a valid attribute to pass, but we don't want to output any svg definition for it. Checking for unset makes sure we're dealing with a valid slug that should output a corresponding svg definition.

scruffian commented:
> It's working in all these cases and only the duotones used on that post are being output in the <svg> footer definitions.

I think there must be something wrong with your testing setup - it shouldn't be working for all these cases yet!

scruffian committed:
save the presets that are used in theme.json for blocks that are actually rendered

scruffian committed:
rename function

scruffian committed:
reability

scruffian commented on lib/block-supports/duotone.php:
```diff
@@ -560,3 +572,88 @@ static function () use ( $filter_svg, $selector ) {
+	}
+);
+
+function gutenberg_save_duotone_preset_svgs( $block_content, $block ) {
```
This runs for every block we render. That might be inefficient...

scruffian committed:
only output the duotone presets that are used

scruffian committed:
remove unused code

scruffian committed:
output custom duotones still

scruffian commented on lib/compat/wordpress-6.2/get-global-styles-and-settings.php:
```diff
@@ -121,46 +121,49 @@ function gutenberg_get_global_stylesheet( $types = array() ) {
-	$styles_variables = '';
-	if ( in_array( 'variables', $types, true ) ) {
+
+	// We need to do this first for duotone.
```
Elucidate this!

scruffian committed:
reset changes to theme json class

scruffian committed:
rename action to filter

scruffian committed:
fixes for classic themes

jeryj committed:
Huge refactor of outputting only duotone css and svg needed for rendered blocks

I'm sorry this is in one giant commit. The general idea is:
- On wp_loaded
  - save all duotone presets defined in merged global styles
  - save all block names and their corresponding duotone slug info
- on block render, check if the block:
  - is one of the blocks using duotone via global styles
  - has any defined duotone styles
  - if so, add it to the duotone_output array to have its duotone info output

jeryj commented:
@scruffian - I branched off of this since I was doing one gigantic refactor and couldn't split it into smaller commits. It is working though (I think)! 🎉  https://github.com/WordPress/gutenberg/pull/48436

If you think it's a good approach, feel free to merge it into this branch and keep going with it.

jeryj commented:
Duotone received a large refactor and has changed too much. Closing in favor of https://github.com/WordPress/gutenberg/pull/48995

scruffian commented:
> Closing in favor of https://github.com/WordPress/gutenberg/pull/48374

Do you mean #48995?

jeryj commented:
haha - yes! Good catch!
